File: app/src/main/java/com/yuriy/openradio/view/EditStationDialog.java
/*
 * Copyright 2017 The "Open Radio" Project. Author: Chernyshov Yuriy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.yuriy.openradio.view;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import com.yuriy.openradio.R;
import com.yuriy.openradio.business.storage.FavoritesStorage;
import com.yuriy.openradio.business.storage.LocalRadioStationsStorage;
import com.yuriy.openradio.utils.AppLogger;
import com.yuriy.openradio.vo.RadioStation;

/**
 * Created by Yuriy Chernyshov
 * At Android Studio
 * On 12/20/14
 * E-Mail: chernyshov.yuriy@gmail.com
 */
public final class EditStationDialog extends BaseAddEditStationDialog {

    /**
     * Tag string to use in logging message.
     */
    private static final String CLASS_NAME = EditStationDialog.class.getSimpleName();

    /**
     * Tag string to use in dialog transactions.
     */
    public static final String DIALOG_TAG = CLASS_NAME + "_DIALOG_TAG";

    public static final String MEDIA_ID_TAG = "MEDIA_ID_TAG";

    private String mMediaId;

    /**
     * Create a new instance of {@link EditStationDialog}
     */
    @SuppressWarnings("all")
    public static EditStationDialog newInstance(final String mediaId) {
        final EditStationDialog editStationDialog = new EditStationDialog();
        final Bundle bundle = new Bundle();
        bundle.putString(MEDIA_ID_TAG, mediaId);
        editStationDialog.setArguments(bundle);
        return editStationDialog;
    }

    @Override
    public View onCreateView(final LayoutInflater inflater, final ViewGroup container,
                             final Bundle savedInstanceState) {
        final View view = super.onCreateView(inflater, container, savedInstanceState);
        getDialog().setTitle(R.string.edit_station_dialog_title);

        final Button addOrEditBtn = view.findViewById(R.id.add_edit_station_dialog_add_btn_view);
        addOrEditBtn.setText(R.string.edit_station_dialog_button_label);

        mMediaId = getMediaId(getArguments());

        if (mMediaId != null) {
            final Context context = getActivity().getApplicationContext();
            final RadioStation radioStation = LocalRadioStationsStorage.getFromLocal(mMediaId, context);
            if (radioStation != null) {
                handleUI(radioStation, context);
            } else {
                //TODO: Handle UI in case of no Radio Station found.
            }
        } else {
            //TODO: Handle UI in case of no ID found.
        }

        return view;
    }

    /**
     * Validate provided input in order to pass data farther to
     * generate {@link RadioStation}.
     *
     * @param name     Name of the Radio Station.
     * @param url      Url of the Stream associated with Radio Station.
     * @param imageUrl Url of the Image associated with Radio Station.
     * @param genre    Genre of the Radio Station.
     * @param country  Country of the Radio Station.
     * @param addToFav Whether or not add radio station to favorites.
     */
    @Override
    protected void processInput(final String name, final String url, final String imageUrl,
                                final String genre, final String country, final boolean addToFav) {
        ((MainActivity) getActivity()).processEditStationCallback(
                mMediaId, name, url, imageUrl, genre, country, addToFav
        );
    }

    /**
     *
     * @param radioStation
     * @param context
     */
    private void handleUI(@NonNull final RadioStation radioStation, final Context context) {
        mNameEdit.setText(radioStation.getName());
        mUrlEdit.setText(radioStation.getStreamURL());
        mImageUrlEdit.setText(radioStation.getImageUrl());
        mCountriesSpinner.setSelection(getCountryPosition(radioStation.getCountry()));
        mGenresSpinner.setSelection(getGenrePosition(radioStation.getGenre()));
        mAddToFavCheckView.setChecked(FavoritesStorage.isFavorite(radioStation, context));
    }

    /**
     *
     * @param bundle
     * @return
     */
    private static String getMediaId(@Nullable final Bundle bundle) {
        if (bundle == null) {
            return null;
        }
        if (!bundle.containsKey(MEDIA_ID_TAG)) {
            return null;
        }
        return bundle.getString(MEDIA_ID_TAG);
    }
}

File: common-ui/src/main/java/com/yuriy/openradio/shared/view/dialog/EditStationDialog.kt
/*
 * Copyright 2017 The "Open Radio" Project. Author: Chernyshov Yuriy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.yuriy.openradio.shared.view.dialog

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import com.yuriy.openradio.shared.R
import com.yuriy.openradio.shared.dependencies.DependencyRegistryCommonUi
import com.yuriy.openradio.shared.utils.AppUtils
import com.yuriy.openradio.shared.utils.findButton
import com.yuriy.openradio.shared.utils.findCheckBox
import com.yuriy.openradio.shared.utils.gone
import com.yuriy.openradio.shared.view.SafeToast.showAnyThread
import com.yuriy.openradio.shared.vo.RadioStation
import com.yuriy.openradio.shared.vo.RadioStationToAdd
import com.yuriy.openradio.shared.vo.getStreamUrl
import com.yuriy.openradio.shared.vo.isInvalid

/**
 * Created by Yuriy Chernyshov
 * At Android Studio
 * On 12/20/14
 * E-Mail: chernyshov.yuriy@gmail.com
 *
 *
 * Dialog to provide components to Edit Radio Station.
 */
class EditStationDialog : BaseAddEditStationDialog() {

    /**
     * Media Id associated with current Radio Station.
     */
    private var mMediaId = AppUtils.EMPTY_STRING

    private lateinit var mEditStationPresenter: EditStationPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        DependencyRegistryCommonUi.inject(this)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = super.onCreateView(inflater, container, savedInstanceState)
        dialog?.setTitle(R.string.edit_station_dialog_title)
        val addOrEditBtn = view!!.findButton(R.id.add_edit_station_dialog_add_btn_view)
        addOrEditBtn.setText(R.string.edit_station_dialog_button_label)
        val addToSrvChkBox = view.findCheckBox(R.id.add_to_srvr_check_view)
        addToSrvChkBox.gone()
        mMediaId = getMediaId(arguments)
        val ctx = requireContext()
        if (mMediaId.isNotEmpty()) {
            val radioStation = mEditStationPresenter.getDeviceLocalRadioStation(mMediaId)
            if (radioStation.isInvalid()) {
                handleInvalidRadioStation(ctx, addOrEditBtn)
            } else {
                handleUI(radioStation)
            }
        } else {
            handleInvalidRadioStation(ctx, addOrEditBtn)
        }
        return view
    }

    /**
     * Validate provided input in order to pass data farther to generate [RadioStation].
     */
    override fun processInput(radioStationToAdd: RadioStationToAdd) {
        val context = activity?.applicationContext ?: return
        mPresenter.editRadioStation(
            context, mMediaId, radioStationToAdd,
            { msg -> onSuccess(msg) },
            { msg -> onFailure(msg) }
        )
    }

    fun configureWith(presenter: EditStationPresenter) {
        mEditStationPresenter = presenter
    }

    /**
     * Handles UI in case of error while trying to edit Radio Station.
     *
     * @param context      Context of a callee.
     * @param addOrEditBtn Edit button.
     */
    private fun handleInvalidRadioStation(context: Context, addOrEditBtn: Button) {
        showAnyThread(context, context.getString(R.string.can_not_edit_station_label))
        addOrEditBtn.isEnabled = false
    }

    /**
     * Update UI with Radio Station loaded from storage.
     *
     * @param radioStation Radio Station.
     */
    private fun handleUI(radioStation: RadioStation) {
        mNameEdit.setText(radioStation.name)
        mUrlEdit.setText(radioStation.getStreamUrl())
        mCountriesSpinner.setSelection(getCountryPosition(radioStation.country))
        mGenresSpinner.setSelection(getGenrePosition(radioStation.genre))
        mAddToFavCheckView.isChecked = mEditStationPresenter.isRadioStationFavorite(radioStation)
    }

    companion object {
        /**
         * Tag string to use in logging message.
         */
        private val CLASS_NAME = EditStationDialog::class.java.simpleName

        /**
         * Tag string to use in dialog transactions.
         */
        val DIALOG_TAG = CLASS_NAME + "_DIALOG_TAG"

        /**
         * Key to keep Media Id's value in Bundle.
         */
        private const val MEDIA_ID_KEY = "MEDIA_ID_KEY"

        fun makeBundle(mediaId: String?): Bundle {
            val bundle = Bundle()
            bundle.putString(MEDIA_ID_KEY, mediaId)
            return bundle
        }

        /**
         * Extract media id from provided Bundle.
         *
         * @param bundle Bundle to handle.
         * @return Media Id or empty string if there is nothing to extract.
         */
        private fun getMediaId(bundle: Bundle?): String {
            if (bundle == null) {
                return AppUtils.EMPTY_STRING
            }
            return if (!bundle.containsKey(MEDIA_ID_KEY)) {
                AppUtils.EMPTY_STRING
            } else bundle.getString(MEDIA_ID_KEY) ?: AppUtils.EMPTY_STRING
        }
    }
}
